Add TPU Pod to doc (#1318)
skypilot-org/skypilot: docs/source/reference/tpu.rst

.. _tpu:

=========
Cloud TPU

> =========

Reviewer: How about we don't add "Overview" to encompass "Free TPUs via TPU Research Cloud (TRC)" and "Getting TPUs in one command"? The reason is these latter 2 headers are eye-catching, so I think they should appear by default in the TOC on the right hand side of the page.

Author: Actually I didn't notice there's a right hand side list. anyways I fixed this.

Overview
========

SkyPilot supports running jobs on Google's `Cloud TPU <https://cloud.google.com/tpu/docs/intro-to-tpu>`_.
Two different TPU architectures are available on GCP:

- `TPU VMs <https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-vm>`_
- `TPU Nodes <https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-node>`_

Both are supported by SkyPilot. We recommend TPU VMs which is a newer architecture encouraged by GCP.

The two architectures differ as follows.
For TPU VMs, you can directly SSH into the "TPU host" VM that is physically connected to the TPU device.

> For TPU Nodes, a user VM (e.g., `n1` instances) must be separately provisioned to communicate with an inaccessible TPU host over gRPC.

Reviewer: ```suggestion
For TPU Nodes, a user VM (an ``n1`` instance) must be separately provisioned to communicate with an inaccessible TPU host over gRPC.
```

Author: I used e.g. because it can be other type of instance rather than `n1`. do you think an is better?

Reviewer: Hmm, I was under the impression that
```
» sky tpunode -t n2d-standard-8
sky.exceptions.ResourcesMismatchError: TPU Nodes can be only used with N1 machines. Please refer to: https://cloud.google.com/compute/docs/general-purpose-machines#n1_machines
```

Author: Ah okay I didn't know this. @WoosukKwon Seems the check was added by you. Is this the case?

.. _tpu:

=========
Cloud TPU
=========

SkyPilot supports running jobs on Google's `Cloud TPU <https://cloud.google.com/tpu/docs/intro-to-tpu>`_.
Two different TPU architectures are available on GCP:

- `TPU VMs <https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-vm>`_
- `TPU Nodes <https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-node>`_

Both are supported by SkyPilot. We recommend TPU VMs which is a newer architecture encouraged by GCP.

The two architectures differ as follows.
For TPU VMs, you can directly SSH into the "TPU host" VM that is physically connected to the TPU device.
For TPU Nodes, a user VM (an `n1` instance) must be separately provisioned to communicate with an inaccessible TPU host over gRPC.

> For more details please refer to GCP `documentation <https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-arch>`_.

Reviewer: How about moving L8-18 to under L45?

Author: good point! just moved.

After the command has finished, you will be dropped into the host VM and can start develop code right away!

Below we demonstrate how to run MNIST training on both TPU VMs and TPU Nodes with SkyPilot YAML.

TPU VMs
--------------------------------

> To use TPU VMs, user needs to specify the TPU type, :code:`tpu_vm: True` and the desired TPU runtime version in :code:`accelerator_args` shown below:

Reviewer: ```suggestion
To use TPU VMs, set the following in a task YAML's ``resources`` field: 
```

.. code-block:: yaml

   resources:
      accelerators: tpu-v2-8
      accelerator_args:

>          runtime_version: tpu-vm-base
>          tpu_vm: True

Reviewer: ```suggestion
         tpu_vm: True
         runtime_version: tpu-vm-base  # optional
```

> 

Reviewer: ```suggestion
The ``accelerators`` field specifies the TPU type, and the :code:`accelerator_args` dict includes the :code:`tpu_vm` bool (defaults to false, which means TPU Node is used), and an optional  TPU ``runtime_version`` field.
```

> Now we show an example of running `mnist training <https://cloud.google.com/tpu/docs/run-calculation-jax#running_jax_code_on_a_tpu_vm>`_ on TPU VM with JAX.

Reviewer: ```suggestion
Here is a complete task YAML that runs `MNIST training <https://cloud.google.com/tpu/docs/run-calculation-jax#running_jax_code_on_a_tpu_vm>`_ on a TPU VM using JAX.
```

.. code-block:: yaml

   name: mnist-tpu-vm

   resources:
      accelerators: tpu-v2-8
      accelerator_args:

>          runtime_version: tpu-vm-base
>          tpu_vm: True

Reviewer: ```suggestion
         tpu_vm: True
         runtime_version: tpu-vm-base
```

   setup: |
      git clone https://github.com/google/flax.git

      conda activate flax
      if [ $? -eq 0 ]; then
         echo 'conda env exists'
      else
         conda create -n flax python=3.8 -y
         conda activate flax
         # Make sure to install TPU related packages in a conda env to avoid package conflicts.
         pip install "jax[tpu]>=0.2.16" -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
         pip install --upgrade clu
         pip install -e flax
      fi

   run: |
      conda activate flax
      cd flax/examples/mnist
      python3 main.py --workdir=/tmp/mnist \
      --config=configs/default.py \
      --config.learning_rate=0.05 \
      --config.num_epochs=10

> A GCS bucket is not required as the TPU VM is physically linked to the TPU device, which can access data directly.

Reviewer: This sentence is a bit out of the blue. Possibly due to reordering?

Author: Fixed. thanks!

> You are expected to see the below outputs when the job finishes.
> 
> .. code-block:: console
> 
>    $ sky launch examples/tpu/tpuvm_mnist.yaml -c mycluster

Reviewer: ```suggestion
This YAML lives under the SkyPilot repo (``examples/tpu/tpuvm_mnist.yaml``), or you can paste it into a local file.

Launch it with

.. code-block:: console

   $ sky launch examples/tpu/tpuvm_mnist.yaml -c mycluster
   
You should see the following outputs when the job finishes.

.. code-block:: console
```

Port some of these to TPU Nodes below?

Author: Fixed, thanks!

   ...
   (mnist-tpu-vm pid=10155) I0823 07:49:25.468526 139641357117440 train.py:146] epoch:  9, train_loss: 0.0120, train_accuracy: 99.64, test_loss: 0.0278, test_accuracy: 99.02
   (mnist-tpu-vm pid=10155) I0823 07:49:26.966874 139641357117440 train.py:146] epoch: 10, train_loss: 0.0095, train_accuracy: 99.73, test_loss: 0.0264, test_accuracy: 99.19

TPU Nodes
--------------------------------

> Different from TPU VM, a host CPU VM needs to be created together with a TPU node and configured correctly to connect with each other.

Reviewer: ```suggestion
In a TPU Node, a normal host CPU VM is created and connected to the TPU device. This is different from a TPU VM.
```

Actually, I don't  get the point about "needs to be created together with a TPU node and configured correctly to connect with each other." -- if you do it using `gcloud` manually or in the GCP console, the user doesn't need to perform this manual effort, right?

Author: What I meant is: user needs to ensure the host VM is created in the same zone as the TPU node and configure `TPU_NAME` env variable. Otherwise they won't be properly connected. But I see the confusion. Not sure how to better express it.

Reviewer: How about we don't mention the complexity. Suggestion: replace L114 with
> In a TPU Node, a normal `n1` CPU VM is provisioned and connected to the TPU host/device.

Then add a new line before L115?

Author: Thanks I updated with slight modification.

> SkyPilot automates the above process with a simple interface:

Reviewer: ```suggestion
To use a TPU Node, set the following in a task YAML's ``resources`` field: 
```

> .. code-block:: yaml

Reviewer: L118-119 change to
```
You can modify the host instance type or the TPU type. To show what TPU types are supported, run :code:`sky show-gpus`.
```

L121 change to:
```
Here is a complete task YAML that runs `MNIST training <https://cloud.google.com/tpu/docs/run-calculation-jax#running_jax_code_on_a_tpu_vm>`_ on a TPU Node using TensorFlow.
```

> After the command has finished, you will be dropped into the host VM and can start develop code right away!

Reviewer: ```suggestion
After the command finishes, you will be dropped into a TPU host VM and can start developing code right away.
```

         pip install --upgrade clu
         pip install -e flax
      fi

   run: |
      conda activate flax
      cd flax/examples/mnist
      python3 main.py --workdir=/tmp/mnist \
      --config=configs/default.py \
      --config.learning_rate=0.05 \
      --config.num_epochs=10

This YAML lives under the `SkyPilot repo <https://github.com/skypilot-org/skypilot/tree/master/examples/tpu>`_ (``examples/tpu/tpuvm_mnist.yaml``), or you can paste it into a local file.

Launch it with:

.. code-block:: console

   $ sky launch examples/tpu/tpuvm_mnist.yaml -c mycluster

You should see the following outputs when the job finishes.

.. code-block:: console

   $ sky launch examples/tpu/tpuvm_mnist.yaml -c mycluster
   ...
   (mnist-tpu-vm pid=10155) I0823 07:49:25.468526 139641357117440 train.py:146] epoch:  9, train_loss: 0.0120, train_accuracy: 99.64, test_loss: 0.0278, test_accuracy: 99.02
   (mnist-tpu-vm pid=10155) I0823 07:49:26.966874 139641357117440 train.py:146] epoch: 10, train_loss: 0.0095, train_accuracy: 99.73, test_loss: 0.0264, test_accuracy: 99.19

TPU Nodes
---------

> In a TPU Node, a normal CPU VM (e.g., `n1` instance) needs to be provisioned to communicate with the TPU host/device.

Reviewer: ```suggestion
In a TPU Node, a normal CPU VM (an `n1` instance) needs to be provisioned to communicate with the TPU host/device.
```

         runtime_version: 2.5.0 # TPU software version to be used.

The above YAML considers :code:`n1-highmem-8` as the host machine and :code:`tpu-v2-8` as the TPU node resource.
You can modify the host instance type or the TPU type.

Here is a complete task YAML that runs `MNIST training <https://cloud.google.com/tpu/docs/run-calculation-jax#running_jax_code_on_a_tpu_vm>`_ on a TPU Node using TensorFlow.

.. code-block:: yaml

> 

Reviewer: After L150, add: `# We use SkyPilot Storage to mount a GCS bucket to /dataset.`

L148 and L132: `<two spaces> # optional, TPU runtime version`

L142, L157, L170: remove? To be consistent with above.

.. note::

   TPU node requires loading data from a GCS bucket. The :code:`file_mounts` spec above simplifies this by using :ref:`SkyPilot Storage <sky-storage>` to create a new bucket/mount an existing bucket.
   If you encounter a bucket :code:`Permission denied` error,
   make sure the bucket is created in the same region as the Host VM/TPU Nodes and IAM permission for Cloud TPU is
   correctly setup (follow instructions `here <https://cloud.google.com/tpu/docs/storage-buckets#using_iam_permissions_for_alternative>`_).

.. note::

>    The environment variable :code:`$TPU_NAME` is automatically set by SkyPilot for connecting TPU devices.

Reviewer: ```suggestion
   The special environment variable :code:`$TPU_NAME` is automatically set by SkyPilot at run time, so it can be used in the ``run`` commands.
```

Author: Updated

>    TPU node requires loading data from a GCS bucket, so we add a :code:`file_mounts` to create a new bucket.

Reviewer: 1. We mount it to /dataset, but it's not used as an argument to mnist_main.py. Expected?
2. It's surprising to see L175-177 being required since we already mount to /dataset? Expected?

Author: Ah actually, mounting locally is not necessary here. TPU device will directly read/write data to the bucket (that's why we need to pass L175-177).
In fact, all I need is just creating a bucket. However, it's neither MOUNT nor COPY so I didn't know which to choose.

Author: But I think mounting could still be useful if user want to do some analysis on the CPU host VM.

>    TPU node requires loading data from a GCS bucket, so we add a :code:`file_mounts` to create a new bucket.
>    Check :ref:`SkyPilot Storage <sky-storage>` for more details.

Reviewer: ```suggestion
   TPU node requires loading data from a GCS bucket. The :code:`file_mounts` spec above simplifies this by using :ref:`SkyPilot Storage <sky-storage>` to create a new bucket/mount an existing bucket.
   
```

Actually, is that creating a new one or mounting an old one?

Author: Either would work right? User can re-use an old one containing processed datasets or create a new one.

TPU Pods
--------------------------------

A `TPU Pod <https://cloud.google.com/tpu/docs/training-on-tpu-pods>`_ is a collection of TPU devices connected by dedicated high-speed network interfaces for scalable training.

> User can simply change the accelerator name to a TPU Pod (e.g., :code:`v2-8` -> :code:`v2-32`) in YAML to launch such resources.

Reviewer: ```suggestion

To use a TPU Pod, simply change the ``accelerators`` field in the task YAML  (e.g., :code:`v2-8` -> :code:`v2-32`).
```

.. code-block:: yaml
   :emphasize-lines: 2-2

   resources:
      accelerators: tpu-v2-32
      accelerator_args:
         runtime_version: tpu-vm-base
         tpu_vm: True

.. note::

>    Both TPU architectures, TPU VMs and TPU Nodes, are supported. Our example is based on TPU VM.

Reviewer: ```suggestion
   Both TPU architectures, TPU VMs and TPU Nodes, can be used with TPU Pods. The example below is based on TPU VMs.
```

To show all available TPU Pods, run :code:`sky show-gpus`:

.. code-block:: console

   GOOGLE_TPU   AVAILABLE_QUANTITIES
   tpu-v2-8     1
   tpu-v2-32    1
   tpu-v2-128   1
   tpu-v2-256   1
   tpu-v2-512   1
   ...

> After creating a TPU Pod, multiple host VMs (e.g., :code:`v2-32` comes with 4 host VMs) will be ready.

Reviewer: This is somewhat confusing. Earlier we explained "host VMs" only appear for TPU Nodes. Is this word sed in the official TPU Pod docs, even for TPU VM+Pod? Should we change it to "worker IPs"?

Author: Yes it's confusing.. But host is the term used in GCP document. 
https://cloud.google.com/tpu/docs/jax-pods

Indeed both TPU Nodes and TPU VMs have the concept of "Host VM" but the former can be paired with any instance type you want and the latter's host instance type is fixed.
Does that clarify it?

TPU Pods
--------------------------------

A `TPU Pod <https://cloud.google.com/tpu/docs/training-on-tpu-pods>`_ is a collection of TPU devices connected by dedicated high-speed network interfaces for scalable training.
To use a TPU Pod, simply change the ``accelerators`` field in the task YAML  (e.g., :code:`v2-8` -> :code:`v2-32`).

.. code-block:: yaml
   :emphasize-lines: 2-2

   resources:
      accelerators: tpu-v2-32
      accelerator_args:
         runtime_version: tpu-vm-base
         tpu_vm: True

.. note::

   Both TPU architectures, TPU VMs and TPU Nodes, can be used with TPU Pods. The example below is based on TPU VMs.

To show all available TPU Pods, run :code:`sky show-gpus`:

.. code-block:: console

   GOOGLE_TPU   AVAILABLE_QUANTITIES
   tpu-v2-8     1
   tpu-v2-32    1
   tpu-v2-128   1
   tpu-v2-256   1

>    tpu-v2-512   1

Author: Fixed.

   ...

After creating a TPU Pod, multiple host VMs (e.g., :code:`v2-32` comes with 4 host VMs) will be ready.
Normally user needs to SSH into all the hosts to setup environments and then launch the job on each host.

> SkyPilot automates such process for you. During :code:`sky launch`, all the setup/run commands will be executed on every host.

Author: Updated with slight modification!

Below we show an YAML example to run a cifar10 training job on :code:`v2-32` TPU Pod with JAX (`code repo <https://github.com/infwinston/tpu-example>`_):

.. code-block:: yaml

   name: cifar-tpu-pod

   resources:
      accelerators: tpu-v2-32
      accelerator_args:
         runtime_version: tpu-vm-base
         tpu_vm: True

   setup: |
      git clone https://github.com/infwinston/tpu-example.git
      cd tpu-example
      pip install "jax[tpu]" -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
      pip install -r requirements.txt

   run: |
      python -u tpu-example/train.py

Expected output from :code:`sky launch`:

.. code-block:: console

>    $ sky launch examples/tpu/cifar_pod.yaml -c mycluster

Author: Good point. updated!

.. code-block:: yaml
   :emphasize-lines: 2-2

   resources:
      accelerators: tpu-v2-32  # Pods have > 8 cores (the last number) 
      accelerator_args:
         runtime_version: tpu-vm-base
         tpu_vm: True

.. note::

   Both TPU architectures, TPU VMs and TPU Nodes, can be used with TPU Pods. The example below is based on TPU VMs.

To show all available TPU Pod types, run :code:`sky show-gpus` (more than 8 cores means Pods):

.. code-block:: console

   GOOGLE_TPU   AVAILABLE_QUANTITIES
   tpu-v2-8     1
   tpu-v2-32    1
   tpu-v2-128   1
   tpu-v2-256   1
   tpu-v2-512   1
   tpu-v3-8     1
   tpu-v3-32    1
   tpu-v3-64    1
   tpu-v3-128   1
   tpu-v3-256   1
   tpu-v3-512   1
   tpu-v3-1024  1
   tpu-v3-2048  1

After creating a TPU Pod, multiple host VMs (e.g., :code:`v2-32` comes with 4 host VMs) are launched.
Normally, the user needs to SSH into all hosts (depending on the architecture used, either the ``n1`` User VMs or the TPU Host VMs) to prepare files and setup environments, and
then launch the job on each host, which is a tedious and error-prone process.

> SkyPilot automates away this complexity. From your laptop, a single :code:`sky launch` command will perform workdir/file syncing, and then execute the setup/run commands on every host of the pod.
> Here is a task YAML for a cifar10 training job on a :code:`v2-32` TPU Pod with JAX (`code repo <https://github.com/infwinston/tpu-example>`_):

Reviewer: ```suggestion
SkyPilot automates away this complexity. From your laptop, a single :code:`sky launch` command will perform:

- workdir/file_mounts syncing; and
- execute the setup/run commands on every host of the pod.

Here is a task YAML for a cifar10 training job on a :code:`v2-32` TPU Pod with JAX (`code repo <https://github.com/infwinston/tpu-example>`_):
```

.. code-block:: yaml

   name: cifar-tpu-pod

   resources:
      accelerators: tpu-v2-32
      accelerator_args:
         runtime_version: tpu-vm-base
         tpu_vm: True

   setup: |
      git clone https://github.com/infwinston/tpu-example.git
      cd tpu-example
      pip install "jax[tpu]" -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
      pip install -r requirements.txt

   run: |
      python -u tpu-example/train.py

Launch it with:

.. code-block:: console

   $ sky launch examples/tpu/cifar_pod.yaml -c mycluster

You should see the following output.

.. code-block:: console

   (node-0 pid=57977, ip=10.164.0.24) JAX process: 1 / 4
   (node-3 pid=57963, ip=10.164.0.26) JAX process: 3 / 4
   (node-2 pid=57922, ip=10.164.0.25) JAX process: 2 / 4
   (node-1 pid=63223) JAX process: 0 / 4
   ...
   (node-0 pid=57977, ip=10.164.0.24) [  1000/100000]      time  0.034 ( 0.063)    data  0.008 ( 0.008)    loss  1.215 ( 1.489)    acc 68.750 (46.163)

.. note::

>    By default, outputs from all hosts will be displayed. You may use :code:`jax.process_index()` to determine which host to print messages.

Reviewer: ```suggestion
   By default, outputs from all hosts are shown with the ``node-<i>`` prefix. Use :code:`jax.process_index()` to control which host to print messages.
```
